Adds hurry up and grace period features to multiball ball save

atummons opened the pull request:
Added hurry up and grace period features to the multiball ball save.  This allows the user to run shows off the hurry up, as well as provide a grace period for saves.

The tests for test_Modes are failing when I run bulk, but run fine when I run just them.  So not sure what is going on there.  If you also see this issue, please let me know what I changed or need to do to fix that issue.

If this is approved and merged, I would also like to go add similar features for add_a_ball to provide a different and shorter ball save feature.

Relates to 768 (set shoot_again to 0) and 1132.  The wording is slightly inconsistent with ball_saves, but I didn't want to change it and break users configs.  But might be worth doing (however that works) to ultimately have them be the same verbiage.

atummons committed:
Adds additional ball save features to multiballs

Added hurry up and grace period features to the multiball ball save.
This allows the user to run shows off the hurry up, as well as provide a
grace period for saves.

atummons committed:
Adds additional ball save features to multiballs

Adds hurry up and grace period

atummons committed:
Added hurry up and grace period features to the multiball ball save.
This allows the user to run shows off the hurry up, as well as provide a
grace period for saves.

The tests for test_Modes are failing when I run bulk, but run fine when I run just them.  So not sure what is going on there.  If you also see this issue, please let me know what I changed or need to do to fix that issue.

If this is approved and merged, I would also like to go add similar features for add_a_ball to provide a different and shorter ball save feature.

atummons committed:
Pleas enter the commit message for your changes. Lines starting

jabdoa2 commented:
Tests fail because of this:

========

mpf/devices/multiball.py

  Line: 147

    pep257: D202 / No blank lines allowed after function docstring (found 1)

  Line: 166

    pep8: E128 / continuation line under-indented for visual indent (col 28)

    pep8: E128 / continuation line under-indented for visual indent (col 28)

atummons committed:
Test of fixing errors

atummons commented:
Jan, Thanks!  I know see where those errors are located in Travis_CI, so I can fix them going forward.  So everything passes now, but on my local it still fails the same 7 tests...  Kinda weird.  They all have the exact same stack trace.  I've tried hunting, but can't seem to find it, and obviously is happening on my machine, but not in Travis.  Thanks!

ERROR: test_ball_end (mpf.tests.test_Modes.TestModes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 548, in setUp
    self._initialise_machine()
  File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 564, in _initialise_machine
    self._wait_for_start(init, 20)
  File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 576, in _wait_for_start
    init.result()
asyncio.base_futures.InvalidStateError: Result is not set.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\desktop\mpf\mpf\tests\MpfTestCase.py", line 557, in setUp
    raise self._exception['exception']
  File "C:\AppData\Local\Programs\Python\Python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "c:\desktop\mpf\mpf\core\events.py", line 115, in _async_handler_done
    future.result()
  File "c:\desktop\mpf\mpf\core\mode_controller.py", line 119, in load_modes
    self.machine.modes[mode] = self._load_mode(mode)
  File "c:\desktop\mpf\mpf\core\mode_controller.py", line 157, in _load_mode
    mode_class = self._load_mode_code(mode_string, config['mode']['code'])
  File "c:\desktop\mpf\mpf\core\mode_controller.py", line 135, in _load_mode_code
    return Util.string_to_class("modes.{}.code.{}".format(mode_string, code_path))
  File "c:\desktop\mpf\mpf\core\utility_functions.py", line 608, in string_to_class
    m = __import__(module)
ModuleNotFoundError: No module named 'modes.mode3'

jabdoa2 commented:
> Jan, Thanks! I know see where those errors are located in Travis_CI, so I can fix them going forward. So everything passes now, but on my local it still fails the same 7 tests... Kinda weird. They all have the exact same stack trace. I've tried hunting, but can't seem to find it, and obviously is happening on my machine, but not in Travis. Thanks!
> ## ERROR: test_ball_end (mpf.tests.test_Modes.TestModes)
> 
> Traceback (most recent call last):
> File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 548, in setUp
> self._initialise_machine()
> File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 564, in _initialise_machine
> self._wait_for_start(init, 20)
> File "c:\users\atummons\desktop\mpf\mpf\tests\MpfTestCase.py", line 576, in _wait_for_start
> init.result()
> asyncio.base_futures.InvalidStateError: Result is not set.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
> File "c:\desktop\mpf\mpf\tests\MpfTestCase.py", line 557, in setUp
> raise self._exception['exception']
> File "C:\AppData\Local\Programs\Python\Python37\lib\asyncio\events.py", line 88, in _run
> self._context.run(self._callback, *self._args)
> File "c:\desktop\mpf\mpf\core\events.py", line 115, in _async_handler_done
> future.result()
> File "c:\desktop\mpf\mpf\core\mode_controller.py", line 119, in load_modes
> self.machine.modes[mode] = self._load_mode(mode)
> File "c:\desktop\mpf\mpf\core\mode_controller.py", line 157, in _load_mode
> mode_class = self._load_mode_code(mode_string, config['mode']['code'])
> File "c:\desktop\mpf\mpf\core\mode_controller.py", line 135, in _load_mode_code
> return Util.string_to_class("modes.{}.code.{}".format(mode_string, code_path))
> File "c:\desktop\mpf\mpf\core\utility_functions.py", line 608, in string_to_class
> m = **import**(module)
> ModuleNotFoundError: No module named 'modes.mode3'

Did you run the tests from inside the mpf folder? I think I fixed something similar recently. I think it needed some __init__.py (which should be optional in python 3 but that seems not to be always true). If CI is happy you at least did not make it worse ;-).

jabdoa2 commented on mpf/devices/multiball.py:
```diff
@@ -147,6 +143,52 @@ def start(self):
+                           ms=(shoot_again_ms +
+                               grace_period_ms),
+                           callback=self.stop)
+            self.delay.add(name='grace_period',
```
I guess those should also be removed in stop and post the event right away if they did not fire before, right? Otherwise, they will fire slightly too late which might lead to weird animations or similar.

jabdoa2 reviewed (commented):
Looks good overall. One small question regarding premature stop of the multiball (i.e. due to mode end).

atummons committed:
Adds mode stop handling for ball save features

atummons committed:
Adds blank line to end of file

jabdoa2 commented on mpf/devices/multiball.py:
```diff
@@ -200,9 +242,19 @@ def stop(self):
+        self.machine.events.remove_handler(self._hurry_up)
         self.machine.events.remove_handler(self._ball_drain_shoot_again)
 
+        self.machine.events.post("multiball_" + self.name + "_grace_period")
```
Now we risk to double post the event. One idea would be to add a boolean class property for each of them. Then set those to true on timer start and to false when then handler posted the event. During stop you could check to see if the event has been posted. Additionally, those could become properties for plsceholders (DeviceMonitor on top of the class).

atummons replied:
Great catch.  I went the placeholder route.  Also added a few other things to only add the handlers if the value was present.  If 0, no point in even posting the event.  So hopefully that will have less events handled, and then only shut it down if true during the stop() method.  If true, it removes the handler, sets it to false, then posts the message.

Thanks for the feedback!  Let me know if you have any other thoughts!

atummons replied:
Added an extra test to confirm this behavior worked for default case as well.

atummons committed:
Adds additional handling for mode end conditions

atummons committed:
Adds additional test case

Adds another test case to ensure that when grace period and hurry up are
not set, that they are never called in an event.

jabdoa2 reviewed (approved):
Looks very good to me. Thanks for making all those changes.
